# 1 - Regression Model

## 1.2 Loading Data
[JSON Viewer](http://jsonviewer.stack.hu)

[Pandas: read_json](https://pandas.pydata.org/docs/reference/api/pandas.read_json.html?highlight=read_json#pandas.read_json)


In [1]:
import pandas as pd
import numpy as np

In [7]:
file_path = 'data/imoveis.json'
df = pd.read_json(file_path, orient='columns')

df

ident  \
0      {'customerID': '775564-BOJSMVON', 'source': 'W...   
1      {'customerID': '660895-AUENKNYY', 'source': 'W...   
2      {'customerID': '751522-JESYFEQL', 'source': 'W...   
3      {'customerID': '714052-GAAEWYKS', 'source': 'W...   
4      {'customerID': '568886-ZIBFOMCC', 'source': 'W...   
...                                                  ...   
73610  {'customerID': '134148-MJIBCMDQ', 'source': 'W...   
73611  {'customerID': '541465-QBQVXXXZ', 'source': 'W...   
73612  {'customerID': '800542-BZREFTQS', 'source': 'W...   
73613  {'customerID': '939419-ECGUNTIH', 'source': 'W...   
73614  {'customerID': '610503-VULDJJZZ', 'source': 'W...   

                                                 listing  
0      {'types': {'unit': 'Outros', 'usage': 'Residen...  
1      {'types': {'unit': 'Apartamento', 'usage': 'Re...  
2      {'types': {'unit': 'Outros', 'usage': 'Residen...  
3      {'types': {'unit': 'Outros', 'usage': 'Comerci...  
4      {'types': {'unit': 'Apartamento', 'usage': 'Re...  
...                                                  ...  
73610  {'types': {'unit': 'Casa', 'usage': 'Residenci...  
73611  {'types': {'unit': 'Casa', 'usage': 'Residenci...  
73612  {'types': {'unit': 'Apartamento', 'usage': 'Re...  
73613  {'types': {'unit': 'Casa', 'usage': 'Residenci...  
73614  {'types': {'unit': 'Apartamento', 'usage': 'Re...  

[73615 rows x 2 columns]

## 1.3 Initial data processing

### Normalization

In [8]:
df1 = pd.json_normalize(df.ident)
df1.head()

customerID   source
0  775564-BOJSMVON  Website
1  660895-AUENKNYY  Website
2  751522-JESYFEQL  Website
3  714052-GAAEWYKS  Website
4  568886-ZIBFOMCC  Website

In [9]:
df2 = pd.json_normalize(df.listing, sep='_')
df2.head()

types_unit  types_usage    address_city  address_location_lon  \
0       Outros  Residencial  Rio de Janeiro            -43.413557   
1  Apartamento  Residencial  Rio de Janeiro            -43.509141   
2       Outros  Residencial  Rio de Janeiro            -43.646786   
3       Outros    Comercial  Rio de Janeiro            -43.330818   
4  Apartamento  Residencial  Rio de Janeiro            -43.480742   

   address_location_lat address_zone      address_neighborhood prices_price  \
0            -22.909429   Zona Oeste                   Taquara        45000   
1            -22.869698   Zona Oeste                Santíssimo        45000   
2            -22.986927   Zona Oeste        Pedra de Guaratiba        50000   
3            -22.881977   Zona Norte                 Cascadura        45000   
4            -23.027653   Zona Oeste  Recreio dos Bandeirantes        50000   

  prices_tax_iptu prices_tax_condo  features_bedrooms  features_bathrooms  \
0               0              150                  0                   0   
1               0              120                  2                   1   
2               0              100                  0                   0   
3             346              468                  0                   1   
4             120              400                  3                   2   

   features_suites  features_parkingSpaces features_usableAreas  \
0                0                       1                   62   
1                0                       1                   44   
2                0                       0                  132   
3                0                       0                   32   
4                1                       1                   60   

  features_totalAreas  features_floors  features_unitsOnTheFloor  \
0                  62                0                         0   
1                   0                0                         0   
2                 132                0                         0   
3                  32                0                         0   
4                   0                3                         4   

   features_unitFloor  
0                   0  
1                   0  
2                   0  
3                   3  
4                   2

### Joining dataframes

In [10]:
data_prop = pd.concat([df1, df2], axis=1)
data_prop.head()

customerID   source   types_unit  types_usage    address_city  \
0  775564-BOJSMVON  Website       Outros  Residencial  Rio de Janeiro   
1  660895-AUENKNYY  Website  Apartamento  Residencial  Rio de Janeiro   
2  751522-JESYFEQL  Website       Outros  Residencial  Rio de Janeiro   
3  714052-GAAEWYKS  Website       Outros    Comercial  Rio de Janeiro   
4  568886-ZIBFOMCC  Website  Apartamento  Residencial  Rio de Janeiro   

   address_location_lon  address_location_lat address_zone  \
0            -43.413557            -22.909429   Zona Oeste   
1            -43.509141            -22.869698   Zona Oeste   
2            -43.646786            -22.986927   Zona Oeste   
3            -43.330818            -22.881977   Zona Norte   
4            -43.480742            -23.027653   Zona Oeste   

       address_neighborhood prices_price  ... prices_tax_condo  \
0                   Taquara        45000  ...              150   
1                Santíssimo        45000  ...              120   
2        Pedra de Guaratiba        50000  ...              100   
3                 Cascadura        45000  ...              468   
4  Recreio dos Bandeirantes        50000  ...              400   

  features_bedrooms  features_bathrooms  features_suites  \
0                 0                   0                0   
1                 2                   1                0   
2                 0                   0                0   
3                 0                   1                0   
4                 3                   2                1   

   features_parkingSpaces  features_usableAreas features_totalAreas  \
0                       1                    62                  62   
1                       1                    44                   0   
2                       0                   132                 132   
3                       0                    32                  32   
4                       1                    60                   0   

  features_floors  features_unitsOnTheFloor  features_unitFloor  
0               0                         0                   0  
1               0                         0                   0  
2               0                         0                   0  
3               0                         0                   3  
4               3                         4                   2  

[5 rows x 21 columns]

In [11]:
data_prop.shape

(73615, 21)

In [12]:
for column in data_prop.columns:
    print('========'*30)
    print(data_prop[column].value_counts())

customerID
775564-BOJSMVON    1
917809-TRNXMOHE    1
125760-XEXDELYV    1
294183-VCUBOXCU    1
465524-CFDRZTHS    1
                  ..
200525-LDRZXCVM    1
938284-QWFBFUXB    1
882696-YVZSJYMI    1
566032-WFPGBXHL    1
610503-VULDJJZZ    1
Name: count, Length: 73615, dtype: int64
source
Website    73615
Name: count, dtype: int64
types_unit
Apartamento    59106
Casa            9389
Outros          5120
Name: count, dtype: int64
types_usage
Residencial    69596
Comercial       4019
Name: count, dtype: int64
address_city
Rio de Janeiro        73607
Queimados                 3
São Gonçalo               2
Cachoeirinha              1
São João de Meriti        1
Armação dos Búzios        1
Name: count, dtype: int64
address_location_lon
-43.365895    399
-43.462590    278
-43.333066    236
-43.383834    232
-43.516473    218
             ... 
-43.231882      1
-43.236952      1
-43.240152      1
-43.185712      1
-43.250934      1
Name: count, Length: 25753, dtype: int64
address_location_lat

## 1.4 Cleaning

### Filter

In [ ]:
filtro = (data_prop['types_usage'] == 'Residencial') & (data_prop['address_city'] ==  'Rio de Janeiro')

#aplicando o filtro 
dados_imoveis = dados_imoveis[filtro]
dados_imoveis.head()

### Variable types

exit `.info`

> 69588 entries, 0 to 73614

We realize that we have to resent the index, so that the information is ordered correctly.
 

Agora está correto.

## 1.5 - Final data processing

### Dealing with null data

#### Data null zone

#### Dados nulos prices


### Change column names and variable selection

# Aula 2 - Regressão linear

## 2.1 - Correlação

*   Correlação: grau da relação entre variáveis. Pode ser positiva ou negativa essa relação. 

*   Multicolinearidade: relação perfeita ou exata entre variáveis.



**Dica de leitura:** [Estatística prática para cientistas de dados: 50 conceitos essenciais](https://www.amazon.com.br/Estat%C3%ADstica-Pr%C3%A1tica-Para-Cientistas-Dados/dp/855080603X/ref=sr_1_21?__mk_pt_BR=%C3%85M%C3%85%C5%BD%C3%95%C3%91&crid=1JJ44C5LJXLR3&keywords=data+science&qid=1657892785&sprefix=data+science%2Caps%2C201&sr=8-21)
e [Estatística: O que é, para que serve e como funciona](https://www.amazon.com.br/Estat%C3%ADstica-para-serve-como-funciona/dp/8537815128/ref=pd_sbs_sccl_3_1/137-9319794-7502061?pd_rd_w=EQoo7&content-id=amzn1.sym.d27ac626-cb3c-4c30-830f-0fd405f2510f&pf_rd_p=d27ac626-cb3c-4c30-830f-0fd405f2510f&pf_rd_r=E896BYW6W206YVVR3SHA&pd_rd_wg=3O0Cg&pd_rd_r=3e3910b0-5f34-4a57-87b2-1d918494d614&pd_rd_i=8537815128&psc=1) [(Naked Statistics: Stripping the Dread from the Data - versão em inglês)](https://www.amazon.com.br/Naked-Statistics-Stripping-Dread-Data/dp/039334777X/ref=sr_1_2?keywords=naked+statistics&qid=1657893464&sprefix=naked+st%2Caps%2C267&sr=8-2&ufe=app_do%3Aamzn1.fos.6a09f7ec-d911-4889-ad70-de8dd83c8a74)

Para mais detalhes consultar [Documentação Seaborn](https://seaborn.pydata.org/generated/seaborn.heatmap.html)

Análise da distribuição da variável resposta.

### Transformação das variáveis 

Estamos realizando uma transformação logarítmica das variáveis para obtermos uma distribuição simétrica (mais próxima da normal), para que possamos usar o modelo de regressão linear. A transformação realizada é:


$$log1p = log(1+x)$$


Lembrar de realizar a transformação inversa para obter o valor original:


$$expm1 = exp(x) - 1$$ 




* Para mais detalhes consultar as documentações [`log1p`](https://numpy.org/doc/stable/reference/generated/numpy.log1p.html), [`expm1`](https://numpy.org/doc/stable/reference/generated/numpy.expm1.html#numpy.expm1) e [`FunctionTransformer`](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.FunctionTransformer.html#sklearn.preprocessing.FunctionTransformer)



### Dummies

## 2.2 - Regressão Linear

* "A análise de regressão diz respeito ao estudo da dependência de uma variável, a variável dependente, em relação a uma ou mais variáveis, as variáveis explanatórias, visando estimar e/ou prever o valor médio (da população) da primeira em termos dos valores conhecidos ou fixados (em amostragens repetidas) das segundas."

* [Documentação Sklearn - LinearRegression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html#sklearn.linear_model.LinearRegression) 

* Regressão Simples: a variável dependente (regressando ou `y`) se relaciona a uma única variável explanatória (regressor ou `X`).

* Regressão Múltipla: a variável dependente (regressando ou `y`) se relaciona a mais de uma única variável explanatória (regressor ou `Xi`).


[slides]

## 2.3 - Ajuste e Previsão

### Divisão dos dados

### Modelo

## 2.4 - Métrica de desempenho

### Coeficiente de Determinação 

* descobrir quão “bem” uma linha de regressão amostral é adequada aos dados
* r2 situa-se entre 0 e 1
* "r2 mede a proporção ou percentual da variação total de Y explicada pelo modelo de regressão."





### Extra

* [Documentação Sklearn - Métricas para Regressão](https://scikit-learn.org/stable/modules/model_evaluation.html#regression-metrics)

* Observação: Estas métricas dependem da escala dos dados!!!

* MSE (Mean Squared Error ou Erro Quadrático Médio): média da diferença **elevada ao quadrado** entre o valor real e o previsto. *(penalidade sobre o erro)*

* MAE (Mean Absolute Error ou Erro Absoluto Médio): média da diferença **absoluta** (módulo) entre o valor real e o previsto.


# Aula 3 - Árvore de Decisão para Regressão






## 3.1 - Árvore de Decisão Classificação versus Regressão

[slides]

[Biblioteca Sklearn - Desicion Tree](https://scikit-learn.org/stable/modules/tree.html#regression)

**Dica de leitura**: [Capítulo 6 - Mãos à obra: aprendizado de máquina com Scikit-Learn, Keras & TensorFlow: Conceitos, ferramentas e técnicas para a construção de sistemas inteligentes](https://www.amazon.com.br/M%C3%A3os-obra-aprendizado-Scikit-Learn-inteligentes/dp/8550815489/ref=sr_1_2?keywords=machine+learning&qid=1658181249&sprefix=machine%2Caps%2C469&sr=8-2&ufe=app_do%3Aamzn1.fos.6d798eae-cadf-45de-946a-f477d47705b9), [Para saber mais: como funciona a árvore de decisão? - Classificação](https://cursos.alura.com.br/course/machine-learning-classificacao-tras-panos/task/107966)



## 3.2 - Implementando o Modelo


**Gráfico de erro de predição**: Um gráfico de erros de predição mostra os alvos reais em relação aos valores previstos. Em um modelo perfeito, esses pontos estariam alinhados em 45 grau.


Para saber mais sobre outras visualizações:
[**Documentação YellowBrick**](https://www.scikit-yb.org/en/latest/)



## 3.3 - Métrica

# Aula 4 - Random Forest para Regressão


## 4.1 - Modelo Ensemble

[slides]

[Biblioteca Sklearn - Random Forest Regressor](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestRegressor.html#sklearn.ensemble.RandomForestRegressor)

**Dica de leitura**: [Artigo Científico - Random Forests](https://link.springer.com/article/10.1023/A:1010933404324), [Wikipédia: Random Forests](https://en.wikipedia.org/wiki/Random_forest#:~:text=Random%20forests%20or%20random%20decision,class%20selected%20by%20most%20trees.), [Capítulo 4 - Livro Introduction to Statistical and Machine Learning Methods for Data Science](https://www.amazon.com.br/dp/B09C2DVZHR/ref=dp-kindle-redirect?_encoding=UTF8&btkr=1), [Capítulo 7 - Mãos à obra: aprendizado de máquina com Scikit-Learn, Keras & TensorFlow: Conceitos, ferramentas e técnicas para a construção de sistemas inteligentes](https://www.amazon.com.br/M%C3%A3os-obra-aprendizado-Scikit-Learn-inteligentes/dp/8550815489/ref=sr_1_2?keywords=machine+learning&qid=1658181249&sprefix=machine%2Caps%2C469&sr=8-2&ufe=app_do%3Aamzn1.fos.6d798eae-cadf-45de-946a-f477d47705b9) 

## 4.2 - Previsão

**Ponto de Atenção**
Os parâmetros são praticamente os mesmos entre o `DecisionTreeRegressor()` e o `RandomForestRegressor()`, com exceção de:

* `n_estimators`: quantidade de árvores na floresta
* `bootstrap`: define se será utilizado o método estatístico de amostragem *bootstrap*
* `oob_score`: define se será utilizado o método de amostragem *out-of-bag* para mensurar os erros da predição
* `n_jobs`: número de *jobs* rodando paralelamente (computação paralela - processadores)
* `warm_start`: faz a adequação de uma nova floresta ou usa uma floresta existente.
* `max_samples`: para o caso do `bootstrap` ser `True`, define o número de amostras a serem extraídas de X para treinar cada árvore base. 

**Gráfico de erro de predição**

## 4.3 - Métrica de desempenho

# Aula 5 - Técnicas de Otimização


## 5.1 - Ferramentas de Otimização

* Por que otimizar?

* Quais ferramentas temos?

[Melhorando os hiperparâmetros - Documentação Sklearn](https://scikit-learn.org/stable/modules/grid_search.html#tuning-the-hyper-parameters-of-an-estimator)

### 1. Cross Validation (Validação Cruzada)

* [Documentação Cross-validation](https://scikit-learn.org/stable/modules/cross_validation.html)

* **Cross Validation** é uma técnica muito utilizada para avaliação de desempenho de modelos e consequente verificação de tendências no aprendizado de máquina. Ele consiste em particionar os dados em subconjuntos (partes), onde um subconjunto é utilizado para treino e outro subconjunto é utilizado para teste e avaliação do desempenho do modelo.

* Os dados particionados são os de treino (70% dos dados originais)!

* **K-fold**: conjunto de treino é dividido de forma aleatória em K subconjuntos de aproximadamente mesmo tamanho, depois são realizados K iterações, onde K-1 subconjuntos são para treino e 1 para validação.

* O desempenho calculado pela validação cruzada k-fold é a média de todos os valores calculados em cada iteração.


### 2. Randomized Parameter Optimization (Otimização Aleatória de Parâmetros)

* [Documentação Randomized Search](https://scikit-learn.org/stable/modules/grid_search.html#randomized-parameter-optimization)

* [Artigo Científico sobre Random Search](https://www.jmlr.org/papers/volume13/bergstra12a/bergstra12a.pdf)

* "Define um espaço de busca como um domínio limitado de valores de hiperparâmetros e pontos de amostra aleatórios nesse domínio."

* Imagem no artigo sobre funcionamento desta busca



## 5.2 - Otimizando a árvore de decisão

### Cross validation


### Random Search Validação

### Random Search Teste

## 5.3 - Otimizando a Random Forest

### Cross validation

### Random Search Validação

Definindo os parâmetros

### Random Search Teste

## 5.4 - Escolhendo o melhor modelo